packages/twenty-front/src/modules/ui/navigation/navigation-drawer/hooks/useWorkspaceSwitching.ts

```diff
@@ -0,0 +1,51 @@
+import { useNavigate } from 'react-router-dom';
+import { useRecoilState, useSetRecoilState } from 'recoil';
+
+import { currentWorkspaceMemberState } from '@/auth/states/currentWorkspaceMemberState';
+import { currentWorkspaceState } from '@/auth/states/currentWorkspaceState';
+import { tokenPairState } from '@/auth/states/tokenPairState';
+import { ColorScheme } from '@/workspace-member/types/WorkspaceMember';
+import { useGenerateJwtMutation } from '~/generated/graphql';
+
+export const useWorkspaceSwitching = () => {
+  const navigate = useNavigate();
+  const setTokenPair = useSetRecoilState(tokenPairState);
+  const [generateJWT] = useGenerateJwtMutation();
+  const [currentWorkspace, setCurrentWorkspace] = useRecoilState(
+    currentWorkspaceState,
+  );
+  const setCurrentWorkspaceMember = useSetRecoilState(
+    currentWorkspaceMemberState,
+  );
+
+  const swtitchWorkspace = async (workspaceId: string) => {
+    if (currentWorkspace?.id === workspaceId) return;
+    const jwt = await generateJWT({
+      variables: {
+        workspaceId,
+      },
+    });
+
+    if (jwt.errors) {
+      throw jwt.errors;
+    }
+
+    if (!jwt.data?.generateJWT) {
+      throw new Error('could not create token');
+    }
+    const { tokens } = jwt.data.generateJWT;
+    setTokenPair(tokens);
+    setCurrentWorkspace(jwt.data.generateJWT.user.defaultWorkspace ?? '');
+    if (jwt.data.generateJWT.user.workspaceMember) {
+      const member = jwt.data.generateJWT.user.workspaceMember;
+      const workspaceMember = {
+        ...member,
+        colorScheme: member.colorScheme as ColorScheme,
+      };
+      setCurrentWorkspaceMember(workspaceMember);
+    }
+    navigate(`/objects/companies`);
```

Comment:
I haven't tested at all (because waiting for backend), curious to see if this will work. I have a feeling we'll have to do more (ex: apollo cache could remain and cause an issue no?)

Reply:
It's safer to do a full reload imo, not try to replicate every function that could be done during initialisation (setWorskpace, setMemeber, etc.)

Reply:
You're correct! The Apollo cache not being updated so I did put a `window.location.reload()` at the end

Reply:
https://github.com/twentyhq/twenty/blob/9035762d432866e2812ac82d2ae5d76a5623167d/packages/twenty-front/src/modules/auth/states/tokenPairState.ts#L6-L10

we are using cookie storage for tokens, so yes returning JWT makes sense after switching workspaces 😄 👍